### PR TITLE
feat: States were added to the windows in a json file, to might know if a window is closed or not

### PR DESCRIPTION
**Pull request by elbuho12Jesus**
...

### CODE REVIEW

File: Source/DataModels/Windows/EditorWindows/EditorWindow.h [hunk 1]
```diff
@@ -3,25 +3,30 @@
...
 	EditorWindow(const std::string& name);
 	virtual void DrawWindowContents() = 0;
 	virtual ImVec2 GetStartingSize() const = 0; 
 
 	ImGuiWindowFlags flags;
+	
 
 private:
 	bool focused;
+	void LoadMeta(const std::string name, bool& enable);
```
Comment (gerard-queralt) [inline Source/DataModels/Windows/EditorWindows/EditorWindow.h:27]: Unused

File: Source/DataModels/Windows/EditorWindows/EditorWindow.h [hunk 2]
```diff
@@ -3,25 +3,30 @@
...
+	void UpdateMeta(bool enabled, const std::string name);
```
Comment (gerard-queralt) [inline Source/DataModels/Windows/EditorWindows/EditorWindow.h:28]: Unused

File: Source/DataModels/Windows/Window.h [hunk 3]
```diff
@@ -7,14 +7,14 @@ class Window
 public:
 	virtual ~Window() {}
 
-	virtual void Draw(bool& enabled) = 0;
+	virtual void Draw(bool &enabled) = 0;
 
 	const std::string& GetName() const;
-	
+	bool enabled;
```
Comment (gerard-queralt) [inline Source/DataModels/Windows/Window.h:13]: Unnecessary

File: Source/DataModels/Windows/WindowMainMenu.cpp [hunk 4]
```diff
@@ -38,21 +43,22 @@ void WindowMainMenu::DrawWindowsMenu()
...
-	}
+	}	
 }
 
 void WindowMainMenu::DrawAbout()
 {
 	if (ImGui::MenuItem("About"))
 	{
-		showAbout = !showAbout;
+		enabled = !showAbout;
```
Comment (gerard-queralt) [inline Source/DataModels/Windows/WindowMainMenu.cpp:59]: This change probably needs to be reverted. Now the about window doesn't work properly.

File: Source/DataModels/Windows/WindowMainMenu.cpp [hunk 5]
```diff
@@ -38,21 +43,22 @@ void WindowMainMenu::DrawWindowsMenu()
...
 	}		
-	about->Draw(showAbout);
+	about->Draw(enabled);
```
Comment (gerard-queralt) [inline Source/DataModels/Windows/WindowMainMenu.cpp:61]:
```suggestion
	about->Draw(showAbout);
```

File: Source/FileSystem/Json.h [hunk 6]
```diff
@@ -15,6 +15,7 @@ class Json
 
 	bool fromBuffer(char*& buffer);
 	void toBuffer(rapidjson::StringBuffer& buffer);
+	void getVectorNames(std::vector<const char*> &vec);
```
Comment (gerard-queralt) [inline Source/FileSystem/Json.h:18]: Why have it as an output parameter and not as a return type? Also, "get" should be capitalized.
```suggestion
	std::vector<std::string> GetVectorNames();
```

File: Source/FileSystem/ModuleResources.cpp [hunk 7]
```diff
@@ -14,6 +14,7 @@
 
 const std::string ModuleResources::assetsFolder = "Assets/";
 const std::string ModuleResources::libraryFolder = "Lib/";
+const std::string ModuleResources::settingsFolder = "Settings/";
```
Comment (gerard-queralt) [inline Source/FileSystem/ModuleResources.cpp:17]: Shouldn't be here, not a resource.

File: Source/FileSystem/ModuleResources.cpp [hunk 8]
```diff
@@ -35,6 +36,7 @@ bool ModuleResources::Start()
 	skyboxImporter = std::make_unique<SkyBoxImporter>();
 
 	CreateAssetAndLibFolders();
+	CreateFolderSettings();
```
Comment (gerard-queralt) [inline Source/FileSystem/ModuleResources.cpp:39]: Shouldn't be here. For now maybe move it to `ModuleEditor`, but we might need to create a "ModuleConfig" or something that handles this.

File: Source/Modules/ModuleEditor.cpp [hunk 9]
```diff
@@ -65,16 +86,26 @@ bool ModuleEditor::Start()
...
 }
 
 bool ModuleEditor::CleanUp()
 {
+	rapidjson::Document doc;
+	Json json(doc, doc);	
+	
+	for (int i = 0; i < windows.size(); ++i) 
+	{
+		windows[i].get()->UpdateState(json);
```
Comment (gerard-queralt) [inline Source/Modules/ModuleEditor.cpp:99]: As I said in one of the first comments, I think we don't need `EditorWindow::UpdateState`. What I would do is write the JSON in `ModuleEditor`, just like reading it. Something like:
```
json[windows[i]->GetName().c_str()] = mainMenu->IsWindowEnabled(i);
```

File: Source/Modules/ModuleEditor.cpp [hunk 10]
```diff
@@ -49,14 +52,32 @@ bool ModuleEditor::Init()
...
 	windows.push_back(std::make_unique<WindowFileBrowser>());
-	windows.push_back(std::make_unique<WindowConsole>());
-	mainMenu = std::make_unique<WindowMainMenu>(windows);
+	windows.push_back(std::make_unique<WindowConsole>());	
+	if(buffer==nullptr)
+	{		
+		rapidjson::StringBuffer newBuffer;
+		for (int i = 0; i < windows.size(); ++i)
+		{
+			json[windows[i].get()->GetName().c_str()]=true;
```
Comment (gerard-queralt) [inline Source/Modules/ModuleEditor.cpp:71]:
```suggestion
			json[windows[i]->GetName().c_str()]=true;
```

File: Source/Modules/ModuleEditor.cpp [hunk 11]
```diff
@@ -65,16 +86,26 @@ bool ModuleEditor::Start()
...
+	rapidjson::Document doc;
+	Json json(doc, doc);	
+	
+	for (int i = 0; i < windows.size(); ++i) 
+	{
+		windows[i].get()->UpdateState(json);		
+	}
+	rapidjson::StringBuffer buffer;
+	json.toBuffer(buffer);
+	std::string lib = "Settings/WindowsStates.cong";
```
Comment (gerard-queralt) [inline Source/Modules/ModuleEditor.cpp:103]:
```suggestion
	std::string settingsPath = "Settings/WindowsStates.conf";
```

File: Source/DataModels/Windows/EditorWindows/EditorWindow.cpp [hunk 12]
```diff
@@ -1,31 +1,65 @@
...
+	this->enabled = enabled;
+}
+
+char* EditorWindow::StateWindows()
+{
+	std::string settingsFolder = "Settings/";
+	char* binaryBuffer = {};
+	if (App->fileSystem->Exists(settingsFolder.c_str()))
+	{		
+		std::string set = "Settings/WindowsStates.cong";
```
[the remark this answers is not in the document]
Comment (gerard-queralt) [reply Source/DataModels/Windows/EditorWindows/EditorWindow.cpp:50]: Also probably better to have it as a static string, like we do Assets and Lib in `ModuleResources`. This way we only need to change it in one place.

**Review by gerard-queralt — CHANGES_REQUESTED** [review]
Left many comments, but mostly minor things. Really great job overall, this will be a great addition!

File: Source/Modules/ModuleEditor.cpp [hunk 13]
```diff
@@ -156,3 +189,27 @@ bool ModuleEditor::IsSceneFocused() const
 {
 	return scene->IsFocused();
 }
+
+char* ModuleEditor::StateWindows()
+{
+	std::string settingsFolder = "Settings/";
```
Comment (gerard-queralt) [inline Source/Modules/ModuleEditor.cpp:195]: This is not needed, since you already have this string as a static member.
```suggestion
```

**Review by gerard-queralt — CHANGES_REQUESTED** [review]
Thank you Jesús, every change you did looks great! You seem to have missed some of my requests hidden in the "Show more", please take a look at those as well :P

**Review by gerard-queralt — APPROVED** [review]

**Review by AuxPort66 — APPROVED** [review]

**Review by gerard-queralt — APPROVED** [review]

**Review by gerard-queralt — APPROVED** [review]

**Review by AuxPort66 — APPROVED** [review]

**Pull request merged**